File: app/src/androidTest/java/feelings/helper/answer/AnswerStoreTest.java
package feelings.helper.answer;

import android.content.Context;
import android.database.Cursor;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;

import org.junit.After;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.threeten.bp.LocalDateTime;

import java.util.ArrayList;
import java.util.List;

import feelings.helper.util.DateTimeUtil;

import static feelings.helper.answer.AnswerContract.COLUMN_NAME_ANSWER;
import static feelings.helper.answer.AnswerContract.COLUMN_NAME_DATE_TIME;
import static feelings.helper.answer.AnswerContract.COLUMN_NAME_QUESTION_ID;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
public class AnswerStoreTest {
    private static final int QUESTION_ID = 1;
    private static final LocalDateTime DATE_TIME = LocalDateTime.now();
    private static final String ANSWER_TEXT = "test answer";

    @BeforeClass
    public static void beforeClass() {
        AnswerStore.deleteAll(InstrumentationRegistry.getTargetContext());
    }

    @After
    public void tearDown() {
        Context context = InstrumentationRegistry.getTargetContext();
        AnswerStore.deleteAll(context);
    }

    @Test
    public void testCreate() {
        Context context = InstrumentationRegistry.getTargetContext();
        Answer answer = new Answer(QUESTION_ID, DATE_TIME, ANSWER_TEXT);

        boolean created = AnswerStore.saveAnswer(context, answer);
        assertTrue(created);
    }

    @Test
    public void testGetAll() {
        Context context = InstrumentationRegistry.getTargetContext();
        AnswerStore.saveAnswer(context, new Answer(QUESTION_ID, DATE_TIME.minusMinutes(3), ANSWER_TEXT));
        AnswerStore.saveAnswer(context, new Answer(2, DATE_TIME.minusMinutes(1), "another one"));
        AnswerStore.saveAnswer(context, new Answer(3, DATE_TIME.minusMinutes(2), "one more"));

        List<Answer> answers = cursorToAnswers(AnswerStore.getAll(context));

        assertEquals(3, answers.size());
        Answer fromDb = answers.get(0);
        assertEquals(2, fromDb.getQuestionId());
        assertEquals(DATE_TIME.minusMinutes(1), fromDb.getDateTime());
        assertEquals("another one", fromDb.getAnswerText());

        fromDb = answers.get(1);
        assertEquals(3, fromDb.getQuestionId());
        assertEquals(DATE_TIME.minusMinutes(2), fromDb.getDateTime());
        assertEquals("one more", fromDb.getAnswerText());

        fromDb = answers.get(2);
        assertEquals(QUESTION_ID, fromDb.getQuestionId());
        assertEquals(DATE_TIME.minusMinutes(3), fromDb.getDateTime());
        assertEquals(ANSWER_TEXT, fromDb.getAnswerText());
    }

    @Test
    public void testGetByQuestionId() {
        Context context = InstrumentationRegistry.getTargetContext();
        AnswerStore.saveAnswer(context, new Answer(QUESTION_ID, DATE_TIME.minusMinutes(3), ANSWER_TEXT));
        AnswerStore.saveAnswer(context, new Answer(2, DATE_TIME.minusMinutes(1), "another one"));
        AnswerStore.saveAnswer(context, new Answer(QUESTION_ID, DATE_TIME.minusMinutes(2), "one more"));

        List<Answer> answers = cursorToAnswers(AnswerStore.getByQuestionId(context, QUESTION_ID));

        assertEquals(2, answers.size());
        Answer fromDb = answers.get(0);
        assertEquals(QUESTION_ID, fromDb.getQuestionId());
        assertEquals(DATE_TIME.minusMinutes(2), fromDb.getDateTime());
        assertEquals("one more", fromDb.getAnswerText());

        fromDb = answers.get(1);
        assertEquals(QUESTION_ID, fromDb.getQuestionId());
        assertEquals(DATE_TIME.minusMinutes(3), fromDb.getDateTime());
        assertEquals(ANSWER_TEXT, fromDb.getAnswerText());
    }

    private List<Answer> cursorToAnswers(Cursor cursor) {
        assertNotNull(cursor);
        try {
            List<Answer> list = new ArrayList<>();
            while (cursor.moveToNext()) {
                int questionId = cursor.getInt(cursor.getColumnIndexOrThrow(COLUMN_NAME_QUESTION_ID));
                LocalDateTime dateTime = LocalDateTime.parse(
                        cursor.getString(cursor.getColumnIndexOrThrow(COLUMN_NAME_DATE_TIME)),
                        DateTimeUtil.DB_FORMATTER);
                String answerText = cursor.getString(cursor.getColumnIndexOrThrow(COLUMN_NAME_ANSWER));
                list.add(new Answer(questionId, dateTime, answerText));
            }
            return list;
        } finally {
            cursor.close();
        }
    }
}

File: app/src/androidTest/kotlin/feelings/guide/data/answer/AnswerStoreTest.kt
package feelings.guide.data.answer

import android.content.Context
import android.database.Cursor
import androidx.test.core.app.ApplicationProvider.getApplicationContext
import com.google.common.truth.Truth.assertThat
import feelings.guide.data.question.Question
import org.junit.After
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import org.threeten.bp.LocalDateTime
import java.util.*

private const val ID: Long = 1
private const val QUESTION_ID: Long = 1
private const val QUESTION_ID_2: Long = 2
private const val ANSWER_TEXT = "test answer"

class AnswerStoreTest {
    private val now = LocalDateTime.now()
    private lateinit var context: Context

    @Before
    fun before() {
        context = getApplicationContext()
    }

    @After
    fun after() {
        AnswerStore.deleteAll(context)
    }

    /**
     * for performance tests only
     */
    //@Test
    fun createForPerformanceTest() {
        for (i in 0..10000) {
            val created = AnswerStore.saveAnswer(context, Answer(QUESTION_ID, now.minusSeconds(i.toLong()), ANSWER_TEXT + i))
            assertThat(created).isTrue()
        }

    }

    @Test
    fun testCreate() {
        // given
        val answer = Answer(QUESTION_ID, now, ANSWER_TEXT)

        // when
        val created = AnswerStore.saveAnswer(context, answer)

        // then
        assertThat(created).isTrue()
        val answers = cursorToAnswers(AnswerStore.getAnswers(context))
        assertThat(answers)
            .containsExactly(answer)
    }

    @Test
    fun testCreateWithId() {
        // given
        val answer = Answer(ID, QUESTION_ID, now, ANSWER_TEXT)

        // when
        val created = AnswerStore.saveAnswer(context, answer)

        // then
        assertThat(created).isTrue()
        val answers = cursorToAnswers(AnswerStore.getAnswers(context))
        assertThat(answers)
            .containsExactly(answer)
    }

    @Test
    fun testEdit() {
        // given
        val answer = Answer(QUESTION_ID, now, ANSWER_TEXT)
        AnswerStore.saveAnswer(context, answer)
        answer.answerText = "updated text"

        // when
        val updated = AnswerStore.updateAnswer(context, answer)

        // then
        assertThat(updated).isTrue()
        val answers = cursorToAnswers(AnswerStore.getAnswers(context))
        assertThat(answers)
            .containsExactly(answer)
    }

    @Test
    fun testGetAll() {
        // given
        val answer1 = Answer(QUESTION_ID, now.minusMinutes(3), ANSWER_TEXT)
        val answer2 = Answer(QUESTION_ID_2, now.minusMinutes(1), "another one")
        val answer3 = Answer(3, now.minusMinutes(2), "one more")
        AnswerStore.saveAnswer(context, answer1)
        AnswerStore.saveAnswer(context, answer2)
        AnswerStore.saveAnswer(context, answer3)

        // when
        val answers = cursorToAnswers(AnswerStore.getAnswers(context))

        // then
        assertThat(answers)
            .containsExactly(answer2, answer3, answer1)
            .inOrder()
    }

    @Test
    fun testGetByQuestionId() {
        // given
        val answer1 = Answer(QUESTION_ID, now.minusMinutes(3), ANSWER_TEXT)
        val answer2 = Answer(QUESTION_ID_2, now.minusMinutes(1), "another one")
        val answer3 = Answer(QUESTION_ID, now.minusMinutes(2), "one more")
        AnswerStore.saveAnswer(context, answer1)
        AnswerStore.saveAnswer(context, answer2)
        AnswerStore.saveAnswer(context, answer3)

        // when
        val answers = cursorToAnswers(AnswerStore.getAnswers(context, QUESTION_ID))

        // then
        assertThat(answers)
            .containsExactly(answer3, answer1)
            .inOrder()
    }

    @Test
    fun testDeleteById() {
        // given
        val answer1 = Answer(QUESTION_ID, now.minusMinutes(3), ANSWER_TEXT)
        val answer2 = Answer(QUESTION_ID_2, now.minusMinutes(1), "another one")
        AnswerStore.saveAnswer(context, answer1)
        AnswerStore.saveAnswer(context, answer2)

        // when
        AnswerStore.deleteById(context, answer1.id)

        // then
        val answers = cursorToAnswers(AnswerStore.getAnswers(context))
        assertThat(answers)
            .containsExactly(answer2)
    }

    @Test
    fun testDeleteByIdAndThenUndo() {
        // given
        val answer = Answer(QUESTION_ID, now, ANSWER_TEXT)
        AnswerStore.saveAnswer(context, answer)
        AnswerStore.deleteById(context, answer.id)

        // when
        AnswerStore.saveAnswer(context, answer)

        // then
        val answers = cursorToAnswers(AnswerStore.getAnswers(context))
        assertThat(answers)
            .containsExactly(answer)
    }

    @Test
    fun testDeleteByQuestionId() {
        // given
        val answer = Answer(QUESTION_ID_2, now.minusMinutes(1), "another one")
        AnswerStore.saveAnswer(context, Answer(QUESTION_ID, now.minusMinutes(3), ANSWER_TEXT))
        AnswerStore.saveAnswer(context, answer)
        AnswerStore.saveAnswer(context, Answer(QUESTION_ID, now.minusMinutes(2), "one more"))

        // when
        AnswerStore.deleteByQuestionId(context, QUESTION_ID)

        // then
        val answers = cursorToAnswers(AnswerStore.getAnswers(context))
        assertThat(answers)
            .containsExactly(answer)
    }

    @Test
    fun testDeleteForDeletedQuestions() {
        // given
        val questionId = QuestionService.createQuestion(context, Question("test question"))
        val answer = Answer(questionId, now.minusMinutes(3), ANSWER_TEXT)
        val answer2 = Answer(QUESTION_ID, now.minusMinutes(2), "one more")
        AnswerStore.saveAnswer(context, answer)
        AnswerStore.saveAnswer(context, answer2)

        // when
        QuestionService.deleteQuestion(context, questionId)
        AnswerStore.deleteForDeletedQuestions(context)

        // then
        val answers = cursorToAnswers(AnswerStore.getAnswers(context))
        assertThat(answers)
            .containsExactly(answer2)
    }

    private fun cursorToAnswers(cursor: Cursor): List<Answer> {
        assertThat(cursor).isNotNull()
        cursor.use {
            val list = ArrayList<Answer>()
            while (it.moveToNext()) {
                list.add(AnswerStore.mapFromCursor(it))
            }
            return list
        }
    }

    companion object {
        @BeforeClass
        @JvmStatic
        fun beforeClass() {
            AnswerStore.deleteAll(getApplicationContext())
        }
    }
}
